stdlib/InteractiveUtils/src/display_precision.jl

```diff
@@ -0,0 +1,178 @@
+# similar to the ones below.
+
+using Printf
+
+########################################################################
+
+export set_display_precision!, unset_display_precision!
+
+# we define global defaults for float display, so that they can be set
+# independent of the interactive display (REPL, IJulia, etc.)
+const DISPLAY_PRECISION = Dict{Type,Union{Int,String}}()
+const DISPLAY_COMPACT_PRECISION = Dict{Type,Union{Int,String}}()
+
+# ensure a valid precision argument
+checkprecision(::Type{T}, precision::Integer) where {T<:AbstractFloat} =
+    precision > 0 || throw(ArgumentError("significant digits = $precision must be positive"))
+checkprecision(::Type{T}, precision::AbstractString) where {T<:AbstractFloat} =
+    Printf.format(Printf.Format(precision), zero(T)) isa AbstractString || throw(ArgumentError("Printf.format failed to return a string"))
+
+function set_display_precision!(::Type{T}, precision::Union{Integer,AbstractString}; compact::Bool=false) where {T<:AbstractFloat}
+    checkprecision(T, precision)
+    foreach(Base.Multimedia.displays) do d
+        set_display_precision!(d, T, precision; compact)
+    end
+    (compact ? DISPLAY_COMPACT_PRECISION : DISPLAY_PRECISION)[T] = precision
+    return
+end
+
+function unset_display_precision!(::Type{T}; compact::Bool=false) where {T<:AbstractFloat}
+    foreach(Base.Multimedia.displays) do d
+        unset_display_precision!(d, T; compact)
+    end
+    delete!(compact ? DISPLAY_COMPACT_PRECISION : DISPLAY_PRECISION, T)
+    return
+end
+
+# type defaults to AbstractFloat:
+set_display_precision!(precision::Union{Integer,AbstractString}; compact::Bool=false) =
+    set_display_precision!(AbstractFloat, precision; compact)
+unset_display_precision!(; compact::Bool=false) =
+    unset_display_precision!(AbstractFloat; compact)
+set_display_precision!(d::AbstractDisplay, precision::Union{Integer,AbstractString}; compact::Bool=false) =
+    set_display_precision!(d, AbstractFloat, precision; compact)
+unset_display_precision!(d::AbstractDisplay, compact::Bool=false) =
+    unset_display_precision!(d, AbstractFloat; compact)
+
+# display subtypes should override these if they store their own precision settings:
+set_display_precision!(d::AbstractDisplay, ::Type, precision::Union{Integer,AbstractString}; compact::Bool=false) = nothing
+unset_display_precision!(d::AbstractDisplay, ::Type; compact::Bool=false) = nothing
+
+########################################################################
+
+struct PrecisionIO{IO_T<:IO} <: Base.AbstractPipe
+    io::IO_T # an underlying IO stream
+
+    precision::Dict{Type,Union{Int,String}} # map from float types to # of sig. digits / format string
+    compact_precision::Dict{Type,Union{Int,String}} # overrides for :compact=>true contexts
+
+    PrecisionIO(io::IO, precision::AbstractDict, compact_precision::AbstractDict) =
+        new{typeof(io)}(io, precision, compact_precision)
+end
```

Comment:
The alternative to defining a new `PrecisionIO <: IO` subtype to propagate this information is to simply use `IOContext` with a `:precision => <digits>` option.   Using a plain `IOContext` has both advantages and disadvantages:

**Advantages** of using `IOContext`:
* Simpler more user-accessible interface.  (Documented with the other `IOContext` flags.)
* Easier for other packages (e.g. DecFP.jl) to support for their own floating-point types (don't have to depend on `InteractiveUtils.PrecisionIO` internals).
* More consistent with our other `IO` metadata (`:compact` etc)
* Easier to use in non-`display` contexts, e.g. if the user wants to write to a CSV file with fewer digits.

**Disadvantages** of using `IOContext`:
* Can only support `%.*g` format, not arbitrary user-defined format strings. (Since otherwise `base/ryu/Ryu.jl` would have to depend on Printf.)
* Less fine-grained control (probably we would just want `precision` to be an integer, apply to all supporting floating-point types, and override `:compact` unconditionally).
* `IOContext` flags are a global namespace, hence precious and poorly composable — it's not clear if any package is already using `:precision` for something else?
* Requires some refactoring of Printf stdlib, to move `%.*g` formatting to `Ryu.jl`.

Overall, I'm starting to be inclined more towards the `IOContext` route.  It's not clear to me that fine-grained control here (different precisions for different types, or `compact` vs non-`compact` mode) is particularly important — people mainly just want a way to show more (or fewer) digits in the REPL.

What do people think?

Reply:
`IOContext` does seem like a more natural way to implement this.